api/http/handler/auth/authenticate_oauth.go

```diff
@@ -57,13 +56,17 @@ func (handler *Handler) validateOAuth(w http.ResponseWriter, r *http.Request) *h
 		return &httperror.HandlerError{http.StatusForbidden, "OAuth authentication is not enabled", errors.New("OAuth authentication is not enabled")}
 	}
 
-	username, err := handler.authenticateOAuth(payload.Code, &settings.OAuthSettings)
+	userData, err := handler.authenticateOAuth(payload.Code, &settings.OAuthSettings)
 	if err != nil {
 		log.Printf("[DEBUG] - OAuth authentication error: %s", err)
 		return &httperror.HandlerError{http.StatusInternalServerError, "Unable to authenticate through OAuth", httperrors.ErrUnauthorized}
 	}
 
-	user, err := handler.DataStore.User().UserByUsername(username)
+	return handler.userProvisioning(w, &userData, settings)
+}
+
+func (handler *Handler) userProvisioning(w http.ResponseWriter, userData *portainer.OAuthUserData, settings *portainer.Settings) *httperror.HandlerError {
```

Comment:
This method is overly large. Code comments within are a good sign that these parts should be extracted into methods. These in turn seem to be very testable from my POV, as they only act on structured data.

Reply:
I splitted the method into several smaller ones. Still unclear if we should directly implement unit tests or if we should bundle them together with the integration tests.

Reply:
:ok_hand: Do you imply that additional test code will lead to a deeper change and thus stronger coupling to the original portainer codebase? That would only be the case if portainer starts to test their code on unit-level. Also, and IMHO, the resulting conflict would not be hard to deal with.

Reply:
Tests are part of the test story for Portainer